Closes #290 ExploreASL_Master contains BIDS Import Option

MichaelStritt opened the pull request:
Hey, so this is my initial implementation. I haven't tested it that extensively yet, but I wanted to show you my ideas on how to do some of these things.

@jan-petr: You suggested to move the import out of the initialization. The new code should do that. To load the data, we potentially have to re-initialize. I implemented that as well. Maybe you could check that out?

@HenkMutsaerts: I had this weird bug, where BIDS2LEGACY converted a bids compliant ASL file to an LEGACY format ASL4D. After that the parseM0 script is called which then says that ASL4D is not a valid BIDS file. This is why I added the try catch statement, which is certainly not the right fix, but I wanted to ask you how we should do this. (**section** `9. Parse M0`)

An additional question is about the DataPar file which is saved at the end of BIDS2LEGACY. Which fields should be stored there? I'm talking about **lines 85-101** in `ExploreASL_ImportWorkflow`.

MichaelStritt commented:
**Note**

> @HenkMutsaerts: I had this weird bug, where BIDS2LEGACY converted a bids compliant ASL file to an LEGACY format ASL4D. After that the parseM0 script is called which then says that ASL4D is not a valid BIDS file. This is why I added the try catch statement, which is certainly not the right fix, but I wanted to ask you how we should do this. (**section** `9. Parse M0`)

Jan Petr fixed this in [PR 456](https://github.com/ExploreASL/ExploreASL/pull/456).

jan-petr commented:
In reaction to the comment above: No need to talk about this bug - I have created this bug by mistake when improving the headers and parameter admin. I have simply added an incorrect parameter check. I have removed that in a different issue (as said above). Due to the parallel fixes in multiple branches, we can now face some errors after merging everything. So I propose the following
1. Remove the try-catch statement - this is not needed anymore (though I understand you needed it in the moment to test everything).
2. We have several branches fixing several errors and it is now possible that separate branches might not work, so testing each branch won't really help that much.
3. So we should merge all the possibly conflicting branches that are there and run a very thorough testing of everything.

MichaelStritt commented:
> In reaction to the comment above: No need to talk about this bug - I have created this bug by mistake when improving the headers and parameter admin. I have simply added an incorrect parameter check. I have removed that in a different issue (as said above). Due to the parallel fixes in multiple branches, we can now face some errors after merging everything. So I propose the following
> 
>     1. Remove the try-catch statement - this is not needed anymore (though I understand you needed it in the moment to test everything).
> 
>     2. We have several branches fixing several errors and it is now possible that separate branches might not work, so testing each branch won't really help that much.
> 
>     3. So we should merge all the possibly conflicting branches that are there and run a very thorough testing of everything.

Totally agree 👍

MichaelStritt commented:
How do you want to do this? Do we merge #456 to develop and then merge develop back to #458?

jan-petr commented on ExploreASL_Master.m:
```diff
@@ -1,37 +1,54 @@
+%ExploreASL_Master ExploreASL pipeline master wrapper calling the individual import & pipeline modules
 %
-% FORMAT: [x] = ExploreASL([DataParPath, ProcessData, SkipPause, iWorker, nWorkers, iModules])
+% FORMAT: [x] = ExploreASL([DataParPath, ImportArray, ProcessArray, SkipPause, iWorker, nWorkers])
```
Good that you keep the parameter names at line four - makes it easier to see what's there.

MichaelStritt replied:
Similar question to the one about the new field for x.StudyRoot. I used x.SkipPause, x.iWorker and x.nWorkers to make the input parsing easier. If that's a problem or you want them in sub-fields of x we can easily do this. Separate variables are just a bit problematic and make the code really ugly there.

jan-petr commented on ExploreASL_Master.m:
```diff
@@ -1,37 +1,54 @@
-function [x] = ExploreASL_Master(DataParPath, ProcessData, SkipPause, iWorker, nWorkers, iModules)
-%ExploreASL_Master ExploreASL pipeline master wrapper calling the individual pipeline modules
+function [x] = ExploreASL_Master(varargin)
```
Seeing that the input parameters changed a bit - have you checked that this change is reflected in all functions that call the master script (ExploreASL_Master or ExploreASL)? There should be at least one, right - xASL_qc_Test...

MichaelStritt replied:
Not yet. I wanted to wait for feedback, so we don't have to change these scripts multiple times. I can do it know if you're happy with the way the Master scripts looks right now.

MichaelStritt replied:
Hopefully fixed all in 8c1269d.

jan-petr commented on ExploreASL_Initialize.m:
```diff
@@ -1,16 +1,10 @@
-function [x] = ExploreASL_Initialize(DataParPath, ProcessData, iWorker, nWorkers)
+function [x] = ExploreASL_Initialize(varargin)
```
Same here - have you changed the calls of ExploreASL_Initialize everywhere if they match to this input parameter change?

MichaelStritt replied:
Same answer as above :)

MichaelStritt replied:
Hopefully fixed all in c70a671.

jan-petr commented on ExploreASL_Initialize.m:
```diff
@@ -1,16 +1,10 @@
-%               - (OPTIONAL, default=prompt the user)
-%   iWorker     - allows parallelization when called externally. iWorker defines which of the parallel ExploreASL calls we are (OPTIONAL, DEFAULT=1)
-%   nWorkers    - allows parallelization when called externally. nWorkers defines how many ExploreASL calls are made in parallel (OPTIONAL, DEFAULT=1)
+%   This script can accept the same arguments as ExploreASL_Master. Check out the definitions there.
```
Good point!

jan-petr replied:
BTW: the guide for good practice in comments and commit messages says "can accept" -> "accepts"

MichaelStritt replied:
Fixed 2cf728a.

jan-petr commented on Functions/xASL_bids_BIDS2Legacy.m:
```diff
@@ -250,17 +250,27 @@ function xASL_bids_BIDS2Legacy(pathStudy, bOverwrite, dataPar)
-if ~isempty(ListASL4D)
-    for iList=1:numel(ListASL4D)
-        xASL_bids_parseM0(ListASL4D{iList});
+try
```
No longer needed. Good that you had it here for testing, but lets remove it now.

MichaelStritt replied:
Removed it here b1758a0.

jan-petr reviewed (changes requested):
Looks all very good. Minor comments only. And crossing fingers that it all works :)

jan-petr commented:
> How do you want to do this? Do we merge #456 to develop and then merge develop back to #458?

1. Merge #456 to develop. Then rebase #458 (as normally done), but run test on the rebased before merging (quick test to make sure that nothing is completely broken).
2. Merge #458 to develop and test thoroughly.

MichaelStritt commented on ExploreASL_ImportWorkflow.m:
```diff
@@ -0,0 +1,123 @@
+    %% Import Workflow
+    
+    % We expect x.DataParPath to be the sourceStructure.json file, so the root directory should be the study directory
+    [x.StudyRoot,~,~] = xASL_fileparts(x.DataParPath);
```
I was almost waiting for feedback on this :)
I just thought it would be a good idea style-wise to create a separate variable for this, but maybe there is already a correct one in the x-struct. `ExploreASL_ImportBIDS` needs the folder containing the sourcedata, rawdata etc. which is the folder of the sourceStructure, which is why I defined this field. We can alternatively just remove it from the x-struct or put it in one of its fields, I guess, just let me know which one.

HenkMutsaerts replied:
It's good to think this a bit through. "DataPar" contained settings about data, this is not allowed in BIDS, and we don't need this anymore as all these settings are now in the json sidecars. So all we need still is a settings par file (which can still be called DataPar for backward compatibility). 

I'm also wondering if we shouldn't restructure the fieldnaming in the x struct while we're at it.
At least we should keep environmental (current choice, potentially platform-dependent) and data parameters (always the same, belong to data) separately. Also Pieter Vandemaele had good ideas about this, and drafted this in his docs on Google Drive

MichaelStritt replied:
Sure, just tell me what you want it to look like. For the parsing it's just quite useful to have all the arguments in the x struct, but I don't care that much about the field. In python people usually have `opts` (options) for the input arguments as a real only structure. The input parser does the same, you get p.results as read only. So I think we could use something like ...
```
x.opts.SkipPause
x.opts.iWorker
x.opts.nWorkers
```
... for the semi-important arguments and then we use ...
```
x.data.StudyRoot
x.data.DataParPath
```
... and maybe a separate processing ...
```
x.proc.ImportArray
x.proc.ProcessArray
```
I just want to make progress, so let me know what you'd like so I can go on implementing.

MichaelStritt replied:
Okay, I added everything to the dev meeting notes. We can discuss this there and update it after the #290 merge in a separate issue.

MichaelStritt commented on ExploreASL_ImportWorkflow.m:
```diff
@@ -0,0 +1,123 @@
+        if strcmp(thisFolderName,'rawdata') && exist(ListFolders{iList},'dir')
+            
+            % Clean the old data
+            if isunix
```
Should be fixed here 593ab7e.  I think this is a relic from the changes during the v1.5.0 release.

HenkMutsaerts commented:
>> @HenkMutsaerts: I had this weird bug, where BIDS2LEGACY converted a bids compliant ASL file to an LEGACY format ASL4D. After that the parseM0 script is called which then says that ASL4D is not a valid BIDS file. This is why I added the try catch statement, which is certainly not the right fix, but I wanted to ask you how we should do this. (section 9. Parse M0)

I need to fix this, so please don't put a try-catch statement here :)

HenkMutsaerts reviewed (changes requested):
Just some questions for now, perhaps you should help me a bit with all these great changes you did ;)

jan-petr reviewed (approved):
Fine to merge. We should go back to the issue of x-struct that wasn't resolved and discuss the new structure properly. But that shouldn't hold merging this. 

The merging and fixing conflicts might be really tricky - especially since some functions were renamed and edited at the same time. Let me know if you want help with that - we can probably reconstruct together all the double-edits.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
 %% -----------------------------------------------------------------------
-%% 3) Print warnings
-fprintf('\n%s\n\n','---------------------------------------------');
+%% Define input parser
```
What is the advantage now of this, if the code only becomes longer with this way of input parsing? ;)
Also, it makes sense to read the input parsing at the first part of the code right. Though it's nice that it's separated here below in a subfunction, it becomes more difficult to debug.

MichaelStritt replied:
Length isn't everything ;)

I think the benefit is that it's modularized now. Of course we could write it in a more simplified way like before, but the use of varargin and the input parser are the most high level solutions described in here: https://docs.google.com/document/d/1M9GTXbRe1VngH9M3p6jyVABqtFgnSYDgpL13MnCelHQ/edit?usp=sharing 

And instead of multiple if-else-statements, you just change the default value or the limiting function handle instead. That seems a lot more sophisticated to me. Why should debugging be harder now?

HenkMutsaerts replied:
That's true. I just thought of debugging somewhere in the script, and usually a script then more or less has a top-down order, where you see at the top which parameters are initialized where and why. So it is faster/easier to see how this works. Now, you have to search for the input parsing function, realize that this is a subfunction, search for the subfunction, and check what's going on there. 

It's also nice to have in 1 screen the function call, header, input parsing, and admin, rather than having the input parsing at the end of the script. Perhaps just intuitively weird. But practice probably will teach me otherwise ;)

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
+    
+    % Define valid input variables
+    validDataParPath = @(variable) ischar(variable) || isempty(variable);
+    validImportArray = @(variable) ischar(variable) || isempty(variable) || isnumeric(variable);
```
Why do all these numerical/logical/Boolean input values allow `ischar(var)`?

MichaelStritt replied:
The compiled version give us char arrays, don't you remember? :)

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
+function parameters = convertParsedInput(parameters)
+
+    % Check if inputs are empty or chars
+    if isempty(parameters.DataParPath),     parameters.DataParPath = '';                                    end
```
Is it confusing if the fieldname is `parameters` here and `x` somewhere else?

MichaelStritt replied:
Hm... I see the x structure as our main structure for everything. So file names, data fields, etc. Here we just parse parameters of ExploreASL_Master/Initialize, so I thought parameters is a good variable.

HenkMutsaerts replied:
True, but you turn `parameters` into `x` in a later stage. Unless fields need to co-exist.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
+function parameters = convertParsedInput(parameters)
+
+    % Check if inputs are empty or chars
+    if isempty(parameters.DataParPath),     parameters.DataParPath = '';                                    end
```
`,` should be `;` though `,` is perhaps also possible, `;` makes more sense right?

MichaelStritt replied:
I always use `,`. Saw both options here https://stackoverflow.com/questions/27561881/one-liner-for-if-then Seems like a style-wise choice.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
-    warning('Subject regexp not specific! Check that no wrong folders are included as subjects');
+%% -----------------------------------------------------------------------
+%% Store parsed input
+function x = storeParsedInput(parameters)
```
Jan and I got the habit to prefix subfunctions by their main function name, so this would be `ExploreASL_Initialize_storeParsedInput` in this case. This makes it nicely readable since this immediately makes it clear that this is a subfunction. If you agree I would apply this for all subfunctions in this function as well (to keep style same throughout)

MichaelStritt replied:
Done in 51f3f3b.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
-fprintf('%s\n',['x.DELETETEMP = ' num2str(x.DELETETEMP) ' (delete temporary files)']);
-fprintf('%s\n',['x.Quality    = ' num2str(x.Quality) ' (0 = fast try-out; 1 = normal high quality)']);
+%% -----------------------------------------------------------------------
+%% Add ExploreASL Directory
```
Very nice that you restructured all this in subfunctions!

MichaelStritt replied:
Thanks, maybe we can even move some to external scripts, to improve the ExploreASL_Initialize readability.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
 
+    % Define paths (should be equal when loading data or only initializing)
+    addpath(MyPath); % ExploreASL
+    subfolders_to_add = {'Functions', 'mex', 'Modules', 'Testing',...
```
perhaps good to order these alphabetically, easier to check

MichaelStritt replied:
Fixed notation to camel case and improved the sorting in 7762591.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
-        fprintf('%s\n',[num2str(size(x.S.SetsID,2)) ' sets are defined for ' num2str(size(x.S.SetsID,1)) ' "SubjectsSessions":']);
+    fprintf('\n%s','Ancillary data, sets: ');
+    if isfield(x.S,'SetsID')
+            fprintf('%s\n',[num2str(size(x.S.SetsID,2)) ' sets are defined for ' num2str(size(x.S.SetsID,1)) ' "SubjectsSessions"']);
```
why did you delete the `:` here?

MichaelStritt replied:
I changed the overall look a little bit. Check out the following example for loading a dataset. If you don't like it, we can change it back.

```
==================================== ExploreASL Settings =====================================
DataParPath		...pment\MATLAB\m.stritt\TestDataSet\analysis\DataParameters_LowQ.json
Run DCM2NII		0
Run NII2BIDS		0
Run ANONYMIZE		0
Run BIDS2LEGACY		0
Run Structural Module	0
Run ASL Module		0
Run Population Module	0
SkipPause		1
iWorker			1
nWorkers		1
==============================================================================================
ExploreASL v1.5.1 initialized ... 
Automatically defining sessions...
==================================== Additional Settings =====================================
1 scans - 0 exclusions, resulting in 1 scans of: 
Longitudinal timePoint 1 = 1 scans - 0 exclusions = 1 scans
ASL sessions: 1

Ancillary data, sets: 3 sets are defined for 1 "SubjectsSessions"
Set 1 = "LongitudinalTimePoint" options "TimePoint_1", codes for paired data
Set 2 = "SubjectNList" options "SubjectNList", codes for paired data
Set 3 = "Site" options "SingleSite", codes for two-sample data

x.D.ROOT		M:\SoftwareDevelopment\MATLAB\m.stritt\TestDataSet\analysis
x.DELETETEMP		1 (delete temporary files)
x.Quality		0 (0 = fast try-out; 1 = normal high quality)

==============================================================================================
```

HenkMutsaerts replied:
Very nice! I would then only make it a bit more compact, by writing all ImportModules on 1 line, and all ProcessModules on 1 line.

Very cool!

MichaelStritt replied:
Sure, if you prefer that. Maybe like this? d358bb6

```
>> [x] = ExploreASL_Master('.\analysis\DataParameters_LowQ.json', 1, 1);
ExploreASL will run the import workflow and will run the processing pipeline...


==============================================================================================
 ________                      __                                 ______    ______   __        
/        |                    /  |                               /      \  /      \ /  |      
########/  __    __   ______  ## |  ______    ______    ______  /######  |/######  |## |      
## |__    /  \  /  | /      \ ## | /      \  /      \  /      \ ## |__## |## \__##/ ## |      
##    |   ##  \/##/ /######  |## |/######  |/######  |/######  |##    ## |##      \ ## |      
#####/     ##  ##<  ## |  ## |## |## |  ## |## |  ##/ ##    ## |######## | ######  |## |      
## |_____  /####  \ ## |__## |## |## \__## |## |      ########/ ## |  ## |/  \__## |## |_____ 
##       |/##/ ##  |##    ##/ ## |##    ##/ ## |      ##       |## |  ## |##    ##/ ##       |
########/ ##/   ##/ #######/  ##/  ######/  ##/        #######/ ##/   ##/  ######/  ########/ 
                    ## |                                                                      
                    ## |                                                                      
                    ##/  

==================================== ExploreASL Settings =====================================
DataParPath		...pment\MATLAB\m.stritt\TestDataSet\analysis\DataParameters_LowQ.json
Import Modules		DCM2NII NII2BIDS ANONYMIZE BIDS2LEGACY
Process Modules		Structural ASL Population
bPause			0
iWorker			1
nWorkers		1
==============================================================================================
ExploreASL v1.5.1 initialized ... 
```

```
>> [x] = ExploreASL_Master('.\analysis\DataParameters_LowQ.json', 0, 0);
ExploreASL will load the dataset...


==============================================================================================
 ________                      __                                 ______    ______   __        
/        |                    /  |                               /      \  /      \ /  |      
########/  __    __   ______  ## |  ______    ______    ______  /######  |/######  |## |      
## |__    /  \  /  | /      \ ## | /      \  /      \  /      \ ## |__## |## \__##/ ## |      
##    |   ##  \/##/ /######  |## |/######  |/######  |/######  |##    ## |##      \ ## |      
#####/     ##  ##<  ## |  ## |## |## |  ## |## |  ##/ ##    ## |######## | ######  |## |      
## |_____  /####  \ ## |__## |## |## \__## |## |      ########/ ## |  ## |/  \__## |## |_____ 
##       |/##/ ##  |##    ##/ ## |##    ##/ ## |      ##       |## |  ## |##    ##/ ##       |
########/ ##/   ##/ #######/  ##/  ######/  ##/        #######/ ##/   ##/  ######/  ########/ 
                    ## |                                                                      
                    ## |                                                                      
                    ##/  

==================================== ExploreASL Settings =====================================
DataParPath		...pment\MATLAB\m.stritt\TestDataSet\analysis\DataParameters_LowQ.json
Import Modules			   
Process Modules			  
bPause			0
iWorker			1
nWorkers		1
==============================================================================================
ExploreASL v1.5.1 initialized ... 
Automatically defining sessions...
==================================== Additional Settings =====================================
1 scans - 0 exclusions, resulting in 1 scans of: 
Longitudinal timePoint 1 = 1 scans - 0 exclusions = 1 scans
ASL sessions: 1

Ancillary data, sets: 3 sets are defined for 1 "SubjectsSessions"
Set 1 = "LongitudinalTimePoint" options "TimePoint_1", codes for paired data
Set 2 = "SubjectNList" options "SubjectNList", codes for paired data
Set 3 = "Site" options "SingleSite", codes for two-sample data

x.D.ROOT		M:\SoftwareDevelopment\MATLAB\m.stritt\TestDataSet\analysis
x.DELETETEMP		1 (delete temporary files)
x.Quality		0 (0 = fast try-out; 1 = normal high quality)

==============================================================================================
```

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
-fprintf('\n%s\n','Ancillary data, sets:');
-if isfield(x.S,'SetsID')
-        fprintf('%s\n',[num2str(size(x.S.SetsID,2)) ' sets are defined for ' num2str(size(x.S.SetsID,1)) ' "SubjectsSessions":']);
+    fprintf('\n%s','Ancillary data, sets: ');
```
Wasn't the `\n` there for a reason?

MichaelStritt replied:
Check out the comment above :)

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -697,111 +639,242 @@
-    x.DELETETEMP = 1;
-%     fprintf('%s\n','Default x.DELETETEMP=1 used (delete files temporarily used for processing)');
-end
+    %xASL_init_PrintCheckSettings Check whether pre-defined settings existed in DATA_PAR.m
```
Personal preference probably, but I don't see the benefit of starting a (sub-)function with a single indent, less will fit on the screen ;)

MichaelStritt replied:
Really? So in python you obviously have to do this, since this is how you write functions there. But I know this from C++ and JavaScript as well. There's even a wiki about this: https://en.wikipedia.org/wiki/Indentation_style I really like this style right here: Variant: 1TBS (OTBS), with this example:

```JavaScript
void checknegative(x) {
    if (x < 0) {
        puts("Negative");
    } else {
        nonnegative(x);
    }
}
```

HenkMutsaerts replied:
Yes, I like this as well in small functions. In longer functions, you don't really use this first indent visually anymore, so it just takes up space. Anyway, this seems indeed a good style to use!

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -27,287 +21,249 @@
 
+    % Recheck the DataParFile, which is possibly not a file or does not exist
+    if ~exist(x.DataParPath,'file') && ~exist(x.DataParPath,'dir') % file = without BIDS Import, dir = BIDS import 
+        fprintf('DataPar file does not exist, ExploreASL will only be initialized...\n');
```
This will not be printed if you don't specify `DataParPath` at all right?
It should still be possible to initialize ExploreASL without getting this output, this output should only there with an invalid `DataParPath` indeed.

MichaelStritt replied:
Oh... here's a mistake in the comment anyway. I'll quickly check it later this afternoon.

MichaelStritt replied:
Should be fixed in 5d88126. I also removed the quite verbose user feedback on default. It's really helpful during debugging, because there are so many different use cases.

MichaelStritt replied:
Damn... I know what you want to do, but it's kind of hard to code. So with the changes in c23eac5, there is no warning that the file does not exist, if you only initialize using `ExploreASL_Master();` or `ExploreASL_Initialize();`, since I check if either the import or the processing is turned on. 
If we turn on processing, then we get the CLI request and if this file doesn't exist either, the warning appears:
```
>> [x] = ExploreASL_Master('I dont exist', 0, 1, 1);
Please insert the path to your DataParFile: 'I dont exist either'
DataPar file does not exist, ExploreASL will only be initialized...
```
If the inserted file exists, the processing will start as usual:
```
>> [x] = ExploreASL_Master('I dont exist', 0, 1, 1);
Please insert the path to your DataParFile: '.\analysis\DataParameters_LowQ.json'
```
This doesn't work for the import workflow right now, meaning we don't ask a second time there:
```
>> [x] = ExploreASL_Master('I dont exist', 1, 0, 1);
DataPar file does not exist, ExploreASL will only be initialized...
```
And if processing and import are turned off, but the file doesn't exist, there is no warning (fallback solution for data loading?):
```
[x] = ExploreASL_Master('I dont exist', 0, 0, 1);
```
But there's the question if we need it there...

HenkMutsaerts replied:
I like this example a lot, but do I now understand correctly that the first argument of `ExploreASL_Master` (i.e. the dataParFile) can be used as original dataPar file (for processing) as well as import-source file (for importing)?
That is perhaps asking for trouble... Unless we always use the "import-source file" for importing and then convert it later to an "original dataPar" file. 

(I know Jan uses a dataPar file for doing the import, and that this can be fine, but we are moving towards having the data parameters in JSON sidecars of the NIfTIs, and the study processing settings in the original dataPar (to make it confusing ;) so perhaps we should at least change the name of the original dataPar to processingPar or something :)

MichaelStritt replied:
That's actually pretty straight forward. We open the json. If the "root" variable is called "x", then it's a DataParFile.

MichaelStritt replied:
Just a bit of additional explanation:
(1) If you run the ImportModules, ExploreASL expects the `sourceStructure.json` file.
(2) If all ImportModules are set to false, ExploreASL expects a `DataPar.json` file.
It's the best way, at least in my opinion. Originally, I gave the Import Modules only the directory and it searched for possible files, but that seems a bit more risky. Jan then told me to pass on the `sourceStructure.json` file and I think it works pretty well.

HenkMutsaerts commented on ExploreASL_Initialize.m:
```diff
@@ -27,287 +21,249 @@
+                if ismac % somehow macOS doesnt show the title of the question dialogue
+                    fprintf('Select folder where ExploreASL is installed\n');
+                end
+                pathstr = uigetdir(CurrCD, 'Select folder where ExploreASL is installed');
```
We decided to get rid of the GUI stuff here right, then I would also always use the `input` here, irregardless of OS.

MichaelStritt replied:
Sure, makes my life easier :)
I thought you only wanted to get rid of the questdlg boxes and not of the uigetdir ones.

MichaelStritt replied:
Removed it: e68fa0d.

HenkMutsaerts commented on ExploreASL_ImportWorkflow.m:
```diff
@@ -0,0 +1,115 @@
+end
+
+%% xASL_Import_BIDS2LEGACY
+function [x] = xASL_Import_BIDS2LEGACY(x)
```
Are you specifying some things that were already specified inside `xASL_bids_BIDS2Legacy`, like the default DataPar fields? I would specify this only once, to facilitate debugging.

MichaelStritt replied:
I'm not sure what you mean. I used the code from `xASL_test_BIDSFlavorsFull` and tried to make it work within this workflow. This needed a few adaptions, which is why we have this subfunction. If there's something specific within the function that you would do differently, please let me know.

HenkMutsaerts replied:
OK let me have a closer look (or please remind me ;)

HenkMutsaerts reviewed (changes requested):
Very very impressed, supernice work!

This of course triggers many new questions from me ;) Hope you can bear with them!


* The default behavior of `ExploreASL_Initialize` is just to initialize ExploreASL, so it should not print text `“Data Par file does not exist, ExploreASL will only be initialized…”`
Same for the statement `“ExploreASL will not run the import workflow …”` Rather it should say something like  `“Initializing ExploreASL…”` This is because `ExploreASL_Initialize` will be called from inside the pipeline or test scripts at multiple locations. This feedback is however nice for the main ExploreASL function, which will be called as full pipeline start.

* Throughout the pipeline where `ExploreASL_Master;` is now used for initialization only (mainly testing/QC functions), I would replace this by `ExploreASL_Initialize`. `ExploreASL_Initialize` should then by default initialize only, and not give any feedback about missing dataPar etc, whereas `ExploreASL_Master` can give this feedback (that's more the start of the fill pipeline).
* `ExploreASL` seems to become redundant, probably easier to remove this and replace by `ExploreASL_Master` (such that we only keep one `ExploreASL` function).
* I would rename `ImportArray` and `ProcessArray` to `ImportModules` and `ProcessModules`. That's more self-explanatory, from the rest of the explanation it's clear that these are arrays, but it should be clear that these arrays specify which modules to run.

When you’re done with all this, I should also check the new output you print on the screen/log. Again, very very nice work! Let me know if you need any help!

MichaelStritt commented on Development/BIDS/ZeroingOut.m:
```diff
@@ -1,8 +1,8 @@
+ExploreASL_Master;
 
-Rdir = 'C:\Users\henkj\Google Drive\WorkGeneral\StudiesProjects\BIDS_ASL\TestDataSet_ImportBIDS\raw'
+Rdir = 'C:\Users\henkj\Google Drive\WorkGeneral\StudiesProjects\BIDS_ASL\TestDataSet_ImportBIDS\raw';
```
Check out this issue right here:  #451. Jan proposes to get rid of this script.

MichaelStritt commented on ExploreASL_ImportConfig.m:
```diff
@@ -34,7 +34,7 @@
 %end
 
 % ExploreASL does not need to be initialized here
```
Ah sorry, I think we used to initialize at this part of the pipeline. Now we always initialize before and we shouldn't recursively call Initialize within Initialize. I can remove that comment ;)

MichaelStritt replied:
Fixed in bb1a0d7.

MichaelStritt commented on ExploreASL_Initialize.m:
```diff
@@ -27,287 +21,249 @@
 
-%% Admin
+    % Define input parser
+    p = inputParsing(varargin{:});
```
There's a comment in the header to check out ExploreASL_Master regarding the input arguments, since ExploreASL_Master & ExploreASL_Initialize share the same input arguments now. I can specify it in both scripts if you prefer that, but I actually prefer declaring them once, otherwise we always have to update both descriptions if there is a change.

HenkMutsaerts replied:
Yes that looked good

MichaelStritt commented on ExploreASL_Master.m:
```diff
@@ -1,37 +1,54 @@
+%                - ASL = Run the ASL Module (BOOLEAN, DEFAULT = 0)
+%                - POPULATION = Run the Population Module (BOOLEAN, DEFAULT = 0)
+%
+%   SkipPause    - TRUE = Pause workflow before ExploreASL pipeline (OPTIONAL, DEFAULT = FALSE)
```
So the default of SkipPause is False, at least for non-compiled mode. Ahhh sorry, It should be FALSE = ...

MichaelStritt replied:
Fixed in 329940e.

MichaelStritt commented on ExploreASL_Master.m:
```diff
@@ -1,37 +1,54 @@
+%
+% EXAMPLES:
+%
+% [x] = ExploreASL_Master('MY_PATH',1,1,1,1,1);
```
👍

MichaelStritt replied:
Fixed in 579fdb2.

MichaelStritt commented on ExploreASL_Master.m:
```diff
@@ -1,37 +1,54 @@
-%                  [1] = only structural, [2] = only ASL, [3] = only
-%                  Population. [1 2] = Structural & ASL module, etc.
-%                  OPTIONAL, DEFAULT = [1 2 3]
+%   DataParPath  - Path to data parameter file (OPTIONAL, DEFAULT = [])
```
👍

MichaelStritt replied:
Fixed in 1fd818d.

MichaelStritt commented on ExploreASL_Master.m:
```diff
@@ -1,37 +1,54 @@
+% [x] = ExploreASL_Master('MY_PATH',[1 1 1 1],[1 1 1],1,1,1);
+% [x] = ExploreASL_Master('MY_PATH','[1 1 1 1]','[1 1 1]','1','1','1');
+%
+% OUTDATED EXAMPLES:
```
Yep, I just thought I'd keep them to recreate similar ones there, but you're right. Will remove them now.

MichaelStritt replied:
Fixed in 579fdb2.

MichaelStritt commented on ExploreASL_Master.m:
```diff
@@ -49,61 +66,27 @@
+    % Initialization
+    x = ExploreASL_Initialize(varargin{:});
     
+    % Import
```
Good idea.

MichaelStritt replied:
Did it similar but a bit different in 579fdb2.

MichaelStritt commented on ExploreASL_Master.m:
```diff
@@ -49,61 +66,27 @@
         end
         return; % skip processing
-    elseif ~isdeployed && ~SkipPause % if this exists, we skip the break here
+    elseif ~isdeployed && ~x.SkipPause % if this exists, we skip the break here
```
👍

MichaelStritt replied:
Changed in 1aa1ce0.

MichaelStritt commented on Functions/xASL_bids_BIDS2Legacy.m:
```diff
@@ -72,15 +72,15 @@ function xASL_bids_BIDS2Legacy(pathStudy, bOverwrite, dataPar)
 
 % Fills in important information in the dataPar if missing
-if ~isfield(dataPar,'x.subject_regexp')
+if ~isfield(dataPar.x,'subject_regexp')
```
I think I know what you mean. We could make super sure that the script doesn't break here using:
```matlab
if ~isfield(dataPar,'x')
    dataPar.x = struct;
end
if ~isfield(dataPar.x,'subject_regexp')
    ...
end
```
Other than that I don't know what you're talking about... Didn't you write the part with the dataPar.x ?

HenkMutsaerts replied:
Haha no then perhaps Jan wrote it?
I'm fine here, I just mean that when somewhere (either reading the dataPar.json or reading an input struct) only have `x.*` rather than `dataPar.x.*`, that it manages this correctly.

E.g., if the `dataPar.json` normally is loaded into `x`, and has `x.subject_regexp`, and now suddenly has `x.dataPar.x.subject_regexp` or `x.x.subject_regexp`

But perhaps this is unnecessary precaution ;)

MichaelStritt replied:
Ahhh! I know what you mean and I think I understand why you think that this should be possible. Earlier we read json files as text files and we built the struct ourselves. Now that we use `spm_jsonread` whenever possible, we get code like this:
```matlab
json = spm_jsonread('file');
```
And if we have a structure like this within the json file...
```json
{"x": [{
	"name": "ExampleDataSet",
	}]}
```
... we always get a variable json with a field called x with a sub-field called name etc.

If you dont want this, we would have to change our json files to something like this:
```json
{
"name": "ExampleDataSet",
"subject_regexp": "...",
...
}
```

... but I don't think that's worth the effort just for code that's a bit shorter ;)

MichaelStritt commented on Functions/xASL_bids_BIDS2Legacy.m:
```diff
@@ -260,7 +260,13 @@ function xASL_bids_BIDS2Legacy(pathStudy, bOverwrite, dataPar)
+
+% Overwrite DataParPath in x structure
+fprintf('Overwriting x.DataParPath...\n');
+dataPar.x.DataParPath = fullfile(pathLegacy, 'DataPar.json');
```
In my head I'd always reinitialize at that point and we basically did exactly the same before, right? We always had `ExploreASL(DataParPath,...)` and if you'd call the pipeline like this, inside the script you'd have an absolute path as well.

HenkMutsaerts replied:
Yes, but we didn't save this path (well we probably did in x.mat but not intentionally :)
Most users like me also just entered

`ExploreASL` (or `ExploreASL_Master`)

which would pop up a GUI dialogue or CLI input dialogue, asking for this DataPar path (but that's basically the same)

MichaelStritt replied:
Hm... okay... Could this be problematic?
Let's try to get the use-cases:
1. A user runs the full pipeline: `[x] = ExploreASL_Master('sourceStructure.json',1,1);`. Here we need to find the DataParFile that was created during BIDS2LEGACY definitely.
2. A user runs the the conventional pipeline: `[x] = ExploreASL_Master('dataParFile.json',0,1);`. In the end the x-struct would contain a field with information about the absolute path of the `dataParFile.json`. If that's a problem we can remove it?
3. A user runs the conventional pipeline on a dataset: `[x] = ExploreASL_Master('dataParFile.json',0,1);`. Then he/she moves the dataset to another drive. He then does a rerun of the ASL module: `[x] = ExploreASL_Master('dataParFile.json',0,[0 1 0]);`. Everything's fine?
Sorry, I really don't get what would be the major problem here... :)

MichaelStritt replied:
Hm... But there's probably a better solution anyway :)

MichaelStritt replied:
Sorry, long story short, we can remove it. We just need the path for the following pipeline steps, but we export it using an output parameter anyway. I haven't tested it extensively yet, but I think this should work: cbf17dd.

MichaelStritt commented on Testing/UnitTests/xASL_ut_UnitTest_function_ExploreASL_Master.m:
```diff
@@ -0,0 +1,269 @@
+if ~exist(fullfile(testPatientDestination,'analysis','Sub1','ASL_1','ASL4D.json'),'file'),      testCondition = false;          end
+if ~exist(fullfile(testPatientDestination,'analysis','Sub1','ASL_1','ASL4D.nii'),'file'),       testCondition = false;          end
+if ~exist(fullfile(testPatientDestination,'analysis','Sub1','ASL_1','M0.json'),'file'),         testCondition = false;          end
+if ~exist(fullfile(testPatientDestination,'analysis','Sub1','ASL_1','M0.nii'),'file'),          testCondition = false;          end
```
I try to use as few ExploreASL functions within the UnitTesting as possible. Otherwise you test functions with functions you need to test. Basically nothing wrong with that, if you test xASL_exist first, but I try to get rid of these scenarios whenever possible.
In addition, I don't want it to find both .nii & .nii.gz here. We want the test to be strict. We always want the same results, not sometimes .nii and sometimes .nii.gz, right?

MichaelStritt commented on ExploreASL_ImportWorkflow.m:
```diff
@@ -0,0 +1,105 @@
+function [x] = ExploreASL_ImportWorkflow(x)
```
No, sounds good to me. I'll rename it.

MichaelStritt replied:
Done in 6fcf5db.

MichaelStritt commented on ExploreASL_Initialize.m:
```diff
@@ -27,287 +21,249 @@
 % EXAMPLE for GUI: ExploreASL_Initialize
-% EXAMPLE for calling externally: ExploreASL_Initialize('//MyDisk/MyStudy/DataPar.m', true, 1, 1);
-% EXAMPLE for debugging/initialization only: [x] = ExploreASL_Initialize('',0);
+% EXAMPLE for calling externally: ExploreASL_Initialize('//MyDisk/MyStudy/DataPar.m');
```
Done in 6fcf5db.

MichaelStritt commented on ExploreASL_Initialize.m:
```diff
@@ -27,287 +21,249 @@
-% EXAMPLE for calling externally: ExploreASL_Initialize('//MyDisk/MyStudy/DataPar.m', true, 1, 1);
-% EXAMPLE for debugging/initialization only: [x] = ExploreASL_Initialize('',0);
+% EXAMPLE for calling externally: ExploreASL_Initialize('//MyDisk/MyStudy/DataPar.m');
+% EXAMPLE for debugging/initialization only: [x] = ExploreASL_Initialize('',0,0);
```
Fixed it in 6fcf5db. If the DataParPath is empty and processing is off, then it won't pause or ask for anything.

MichaelStritt commented:
> The default behavior of `ExploreASL_Initialize` is just to initialize ExploreASL, so it should not print text `“Data Par file does not exist, ExploreASL will only be initialized…”`
> Same for the statement `“ExploreASL will not run the import workflow …”` Rather it should say something like  `“Initializing ExploreASL…”` This is because `ExploreASL_Initialize` will be called from inside the pipeline or test scripts at multiple locations. This feedback is however nice for the main ExploreASL function, which will be called as full pipeline start.

I know what you mean. The problem was that there are a lot of different use-cases. What happens if we call ExploreASL with a sourceStructure file, what if it's a DataPar file, what happens if the user only wants to load the data, what happens if the DataPar is invalid but the user set processing to true, ... etc. I've put the quite verbose statements into an if section, so that we can turn it on for debugging purposes. Check out 5d88126.

> Throughout the pipeline where `ExploreASL_Master;` is now used for initialization only (mainly testing/QC functions), I would replace this by `ExploreASL_Initialize`. `ExploreASL_Initialize` should then by default initialize only, and not give any feedback about missing dataPar etc, whereas `ExploreASL_Master` can give this feedback (that's more the start of the fill pipeline).

Definitely agree. Changed this in 33865d3.

> `ExploreASL` seems to become redundant, probably easier to remove this and replace by `ExploreASL_Master` (such that we only keep one `ExploreASL` function).

Sure, also within this PR?

> I would rename `ImportArray` and `ProcessArray` to `ImportModules` and `ProcessModules`. That's more self-explanatory, from the rest of the explanation it's clear that these are arrays, but it should be clear that these arrays specify which modules to run.

Good idea. I changed this in 14e9d8b. Hope I got all the different places in the code, but we'll have to extensively test anyway.

HenkMutsaerts commented:
https://github.com/ExploreASL/ExploreASL/commit/5d8812659ac96f95b8199baf8b746068a6319fd8
-> I responded here!

>> ExploreASL seems to become redundant, probably easier to remove this and replace by ExploreASL_Master (such that we only keep one ExploreASL function).

> Sure, also within this PR?

Haha, well you choose ;) To the user it is probably the same effort, in both functions they switch the usage. The use of `ExploreASL` as addition to `ExploreASL_Master` becomes a bit irrelevant because of this PR, that's why.

HenkMutsaerts commented on ExploreASL_Master.m:
```diff
@@ -17,7 +17,7 @@
 %                - POPULATION = Run the Population Module (BOOLEAN, DEFAULT = 0)
 %
-%   SkipPause    - TRUE = Pause workflow before ExploreASL pipeline (OPTIONAL, DEFAULT = FALSE)
+%   SkipPause    - FALSE = Pause workflow before ExploreASL pipeline (OPTIONAL, DEFAULT = FALSE)
```
Haha, this is my fault. I guess the name of a Boolean should always explain a positive thing.
So this Boolean should be called `bPause`. I called this Boolean `bSkipPause` as you by default want to pause (from the perspective that the user starts ExploreASL).

Let's rename this parameter to `bPause`, that is defaulted to FALSE (no pause). 
But in some situations we might want to default to bPause = TRUE, to give the user the chance to look at the screen settings before continuing. Or shall we just remove this parameter at all, and instruct the "advanced user" (the one that won't use Maurice's GUI) to start running ExploreASL using the "load data only" option (as a sort of dummy start of ExploreASL)?

MichaelStritt replied:
I'm fine with both. But using false on default already sounds like a really good idea ;) Should be done in 3c2529f.

MichaelStritt commented:
> * The first result of `ExploreASL_Initialize` (i.e. what it will always do) is the initialization.
> * I think it's always good to print what you're doing, my comment was just about printing what ExploreASL is NOT doing (e.g., "initializing ExploreASL but not importing BIDS") as this NOT statement can confuse if you're just initializing ExploreASL for some tests and suddenly you get a message about BIDS :)
> * So my solution would be: when running `ExploreASL_Initialize`, always print `Initializing ExploreASL`, and likewise, `run the processing pipeline` and `run the import workflow` I would always print if this is the case.
> * So it seems that I would set `debugFeedback = 1` and disable only the line of `x.ImportData==0` (as this states something we do NOT do).

Sounds good 👍  821f19f should do that now.

MichaelStritt commented:
> This looks good, then we could even make this a warning right?
> Since the user asks for importing or processing data and there is no data par file.
> Shouldn't it then be:
> `if any(x.ImportData>0) || any(x.ProcessData>0)`
> ?

Currently I'm using sum instead of any, to check if one of the values is set to true. Any basically does the same at this point as far as I can see.

MichaelStritt commented:
> Should `x.ImportData` be `x.bImportData` ?

Hm... I like it like it is right now. But if you want we can change it. I'd prioritize finding good fields to put these variables in instead.

MichaelStritt commented:
@HenkMutsaerts: I think 99% is fine now. If you're not totally unhappy, it would be cool if Jan & I could merge this tomorrow. We then test it and improve it (fields for the variable groups, additional warnings etc.). Testing it would be way much easier if I'd have some of the other recently closed PRs. I don't want to rebase even more, since we already expect quite a few conflicts. Please? :)

HenkMutsaerts commented:
> > This looks good, then we could even make this a warning right?
> > Since the user asks for importing or processing data and there is no data par file.
> > Shouldn't it then be:
> > `if any(x.ImportData>0) || any(x.ProcessData>0)`
> > ?
> 
> Currently I'm using `sum `instead of `any`, to check if one of the values is set to true. `any `basically does the same at this point as far as I can see.

Sure, but it looked as if you used something like `if x.ImportData || x.ProcessData`. 

> > Should `x.ImportData` be `x.bImportData` ?
> 
> Hm... I like it like it is right now. But if you want we can change it. I'd prioritize finding good fields to put these variables in instead.

I only meant prefixing variables that are Booleans by the letter "b" :)

MichaelStritt commented:
> 
> 
> > > This looks good, then we could even make this a warning right?
> > > Since the user asks for importing or processing data and there is no data par file.
> > > Shouldn't it then be:
> > > `if any(x.ImportData>0) || any(x.ProcessData>0)`
> > > ?
> > 
> > 
> > Currently I'm using `sum `instead of `any`, to check if one of the values is set to true. `any `basically does the same at this point as far as I can see.
> 
> Sure, but it looked as if you used something like `if x.ImportData || x.ProcessData`. I don't care for the difference between `sum` and `any`, perhaps `any` allows combining logical and numerical format? Otherwise both should be fine indeed.
> 
> > > Should `x.ImportData` be `x.bImportData` ?
> > 
> > 
> > Hm... I like it like it is right now. But if you want we can change it. I'd prioritize finding good fields to put these variables in instead.
> 
> I only meant prefixing variables that are Booleans by the letter "b" :)

Oh, I understand, makes sense. I changed that in bc39716 .
